src/psyclone/psyir/transformations/arrayassignment2loops_trans.py

```diff
@@ -368,6 +367,34 @@ def validate(self, node, options=None):
                 # pylint: disable=cell-var-from-loop
                 raise TransformationError(LazyString(
                     lambda: f"{message} In:\n{node.debug_string()}"))
+            # We must understand any array accesses on the RHS. This means all
+            # array indices must be known (because otherwise they themselves
+            # might be arrays, e.g.
+            #      a(:) = b(c) where `c` can be a scalar or an array.
+            if isinstance(reference, ArrayMixin):
+                # Check the index expressions
+                range_count = len([x for x in accessor.indices
+                                   if isinstance(x, Range)])
+                if range_count != num_of_ranges:
+                    # If the number of ranges in this access is not the same as
+                    # on the LHS then we may or may not have a scalar.
```

Comment:
Can you add: "To disambiguate we need to know that the Reference2ArrayRangeTrans in the apply will provide the remaining ranges, and for this it will need to know the type of each index expression"? (if this is the right reasoning here)

Reply:
I'm not sure. I've extended the comment - see what you think.